### PR TITLE
Contributing update review tips

### PR DESCRIPTION
**Pull request by acarbonetto**
### Description
Update contributing page to include tips/tricks to help developers get their PRs into an approved state (as discussed here: https://opensearch.slack.com/archives/C038PDJ14JZ/p1681232261749639?thread_ts=1681231289.117129&cid=C038PDJ14JZ)

### Issues Resolved
none

### Check List
- [N/A] New functionality includes testing.
  - [N/A] All tests pass
- [N/A] New functionality has been documented.
  - [N/A] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [N/A] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -162,3 +162,18 @@ During the PR process, expect that there will be some back-and-forth. Please try
 If we accept the PR, a [maintainer](MAINTAINERS.md) will merge your change and usually take care of backporting it to appropriate branches ourselves.
 
 If we reject the PR, we will close the pull request with a comment explaining why. This decision isn't always final: if you feel we have misunderstood your intended change or otherwise think that we should reconsider then please continue the conversation with a comment on the PR and we'll do our best to address any further points you raise.
+
+We have a lot of mechanisms to help expedite towards an accepted PR.  Here are some tips for success:
+1. *Minimize BWC guarantees*: First PR review heavily focuses on the public facing API. This is what we have to "guarantee" as non-breaking for bwc across major versions.
+2. *Do not copy ES code*: SSPL co-mingling will get rejected real fast.
```
Comment (dblock) [inline CONTRIBUTING.md:168]: This is incorrect. 

You're not allowed to look at/copy non-APLv2 compatible code, which includes SSPL. You are actually allowed to copy code from ES < 7.10.2 which was APLv2 for example.

So let's stay out of calling out ES explicitly and replace this with something like ensure that code is APLv2 compatible and confirm that you have not copied any code that's not compatible if it implements a feature that may or may not exist in another product?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -162,3 +162,18 @@ During the PR process, expect that there will be some back-and-forth. Please try
...
+6. *Micro-benchmark critical path*: This is a lesser known mechanism, but if you have critical path changes you're afraid will impact performance (gc, heap, direct memory, CPU) then including a [microbenchmark](https://github.com/opensearch-project/OpenSearch/tree/main/benchmarks) with your PR (and jfr or flamegraph results in the description) is a *GREAT IDEA* and will help expedite the review process.
+7. *test, test, test*: pretty self explanatory ([OpenSearchTestCase](./test/framework/src/main/java/org/opensearch/test/OpenSearchTestCase.java) for unit tests, [OpenSearchIntegTestCase](./test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java) for integration / cluster tests, [OpenSearchRestTestCase](./test/framework/src/main/java/org/opensearch/test/rest/OpenSearchRestTestCase.java) for testing REST endpoint interfaces, and yaml tests with [ClientYamlTestSuiteIT](./rest-api-spec/src/yamlRestTest/java/org/opensearch/test/rest/ClientYamlTestSuiteIT.java) for REST integration tests)
```
Comment (dblock) [inline CONTRIBUTING.md:176]: Nit: Get rid of spaces before/after `/` to make it consistent with other text. Same for `w/`?

**Review by dblock — CHANGES_REQUESTED** [review]
I like it. I think the ES mention needs fixing because it's incorrect.

**Review by dblock — COMMENTED** [review]
Should this be in https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md actually?

**Comment by dblock** [comment]
@acarbonetto Yeah I think that would be best because we copy CONTRIBUTING.md for general high level stuff across all repos. You'll get it read more often if it's int he developer guide. Sorry to create more work for you!

**Comment by kotwanikunal** [comment]
Apologies. This PR was auto closed without reaching a resolution from the maintainers.
Re-opening to move it forward.
Thanks for your contributions to OpenSearch!

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -162,3 +162,18 @@ During the PR process, expect that there will be some back-and-forth. Please try
 If we accept the PR, a [maintainer](MAINTAINERS.md) will merge your change and usually take care of backporting it to appropriate branches ourselves.
 
 If we reject the PR, we will close the pull request with a comment explaining why. This decision isn't always final: if you feel we have misunderstood your intended change or otherwise think that we should reconsider then please continue the conversation with a comment on the PR and we'll do our best to address any further points you raise.
+
+We have a lot of mechanisms to help expedite towards an accepted PR.  Here are some tips for success:
+1. *Minimize BWC guarantees*: First PR review heavily focuses on the public facing API. This is what we have to "guarantee" as non-breaking for bwc across major versions.
```
Comment (BhumikaSaini-Amazon) [inline CONTRIBUTING.md:167]: We should also link https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#backwards-compatibility here for details on this front

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -162,3 +162,18 @@ During the PR process, expect that there will be some back-and-forth. Please try
...
+7. *test, test, test*: pretty self explanatory ([OpenSearchTestCase](./test/framework/src/main/java/org/opensearch/test/OpenSearchTestCase.java) for unit tests, [OpenSearchIntegTestCase](./test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java) for integration & cluster tests, [OpenSearchRestTestCase](./test/framework/src/main/java/org/opensearch/test/rest/OpenSearchRestTestCase.java) for testing REST endpoint interfaces, and yaml tests with [ClientYamlTestSuiteIT](./rest-api-spec/src/yamlRestTest/java/org/opensearch/test/rest/ClientYamlTestSuiteIT.java) for REST integration tests)
+
+If you're bias towards liberal guardrails, you have a higher chance of the PR getting merged quickly. We can always relax these guard rails in smaller followup PRs. Reverting a GA feature is much more difficult. Check out the [DEVELOPER_GUIDE](./DEVELOPER_GUIDE.md#submitting-changes) for more useful tips.
```
Comment (BhumikaSaini-Amazon) [inline CONTRIBUTING.md:178]: > If you're bias towards liberal guardrails ...

Should this be "If you are biased towards restrictive guardrails ..." instead?

**Comment by acarbonetto** [comment]
> Hi @acarbonetto, Is this being worked upon? Feel free to reach out to maintainers for further reviews.

Hi @ticheng-aws - sorry I've been off project for a while.  I'll take a look in day or two if nobody else does.

**Comment by scrawfor99** [comment]
I opened a PR to get this over the finish line. Thanks for all the work @acarbonetto. At this point we can close this. @peternied could you help there? Thanks

**Comment by kotwanikunal** [comment]
I can close this one @scrawfor99! Please refer to #12981 for further updates.

**Pull request closed without merging**